core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerBackend.scala

```diff
@@ -344,12 +354,20 @@ private[spark] class MesosSchedulerBackend(
 
   override def frameworkMessage(d: SchedulerDriver, e: ExecutorID, s: SlaveID, b: Array[Byte]) {}
 
+  /**
+   * Remove executor associated with slaveId in a thread safe manner.
+   */
+  private def removeExecutor(slaveId: String) = {
+    synchronized {
```

Comment:
In MesosSchedulerBackend there shouldn't be any need for synchornization. Mesos scheduler driver is built on libprocess and it gurantees one thread executing a message at a time.
Coarse mesos scheduler backend needs to synchronize since the instance is shared.